### PR TITLE
AgentTestRunner refactor and cleanup

### PR DESCRIPTION
**Pull request by randomanderson**
The previous iteration of `AgentTestRunner` had numerous issues which made #2213 impossible:

- Abuse of static fields made it hard to reason about and debug.  State was shared in unintuitive ways between tests
- The `CoreTracer` instance was not only shared between tests in the same class but also between *multiple test classes*
- Abuse of `@BeforeClass`/`@AfterClass` on non-static methods made it difficult to follow or predict test execution
- The bytebuddy listener (`TEST_LISTENER`) jumped through multiple hoops to connect static<->nonstatic instances

### Thus, I made the following changes:

- **Convert AgentTestRunner to a groovy class**: 
This allows the use of `@Shared`, `setupSpec()`, `setup()`, `cleanup()` and `cleanupSpec()`.  All actions take place in clearly defined steps.

- **Removal of all static field state**
For obvious reasons.

- **Add `forceInstallGlobalTracer` to `TracerInstaller`**: 
Previously, whichever test class accessed the static block _**first**_ in `AgentTestRunner` set the tracer instance for all tests on the classloader.  Adding this method allows setting and resetting the tracer per test class.

- **Make `AgentTestRunner` implement `AgentBuilder.Listener` directly**:
`TEST_LISTENER` was clunky.  Now, derived test classes that want to modify transformation behavior can do so with a simple override.

- **Rework of `WithGlobalTracer`**:
The log context tests broke with the above changes because `WithGlobalTracer` made the previously correct assumption that a tracer is only installed once.  With this PR, a tracer is installed for each test class and the callbacks need to be called again. Given that installing a tracer is rare (once per JVM outside of tests) and adding a callback is rare (once per logging framework), I simplified the code by just using `synchronized` blocks where appropriate.

Things I did not do:

- **Update field names**
Changing `TEST_WRITER` to `testWriter` would muddy up the PR for little benefit

- **Create a new `CoreTracer` per test**
This should work now but I'm leaving that for a followup PR.  Leaving the status quo of `CoreTracer` per test class

The end result is a much cleaner (and ~120 lines shorter) base test class

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/TracerInstaller.java [hunk 1]
```diff
@@ -25,9 +25,21 @@ public static void installGlobalTracer(final CoreTracer tracer) {
...
       GlobalTracer.registerIfAbsent(tracer);
       AgentTracer.registerIfAbsent(tracer);
+
+      log.debug("Global tracer installed");
+    } catch (final RuntimeException re) {
+      log.warn("Failed to register tracer: {}", tracer, re);
+    }
+  }
+
+  public static void forceInstallGlobalTracer(CoreTracer tracer) {
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/TracerInstaller.java:35]: Consider adding a warning level log message when this is used so it can be identified if somehow invoked in non-test env's.

File: dd-trace-api/src/main/java/datadog/trace/api/GlobalTracer.java [hunk 2]
```diff
@@ -31,50 +32,52 @@ public boolean addTraceInterceptor(TraceInterceptor traceInterceptor) {
         @Override
         public void addScopeListener(ScopeListener listener) {}
       };
-  private static final AtomicReference<Tracer> provider = new AtomicReference<>(NO_OP);
+
+  private static final Collection<Callback> installationCallbacks = new ArrayList<>();
```
Comment (tylerbenson) [inline dd-trace-api/src/main/java/datadog/trace/api/GlobalTracer.java:36]: Do we need a way for tests to remove callbacks after test completion?

File: dd-java-agent/testing/src/test/groovy/context/FieldBackedProviderTest.groovy [hunk 3]
```diff
@@ -32,21 +33,14 @@ import static datadog.trace.bootstrap.config.provider.SystemPropertiesConfigSour
...
+  void onDiscovery(String typeName, ClassLoader classLoader, JavaModule module, boolean loaded) {
+    if (typeName?.endsWith("UntransformableKeyClass")) {
+      throw new AbortTransformationException(
+        "Aborting transform for class name = " + typeName + ", loader = " + classLoader)
+    }
 
-  @Override
-  protected boolean shouldTransformClass(final String className, final ClassLoader classLoader) {
-    return className == null || (!className.endsWith("UntransformableKeyClass"))
+    super.onDiscovery(typeName, classLoader, module, loaded)
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/test/groovy/context/FieldBackedProviderTest.groovy:43]: I'm not sure I understand this change... please elaborate.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+ * <ul>
+ *   <li>All {@link Instrumenter}s on the test classpath will be applied. Matching preloaded classes
+ *       will be retransformed.
+ *   <li>{@link AgentTestRunner#TEST_WRITER} will be registered with the global tracer and available
+ *       in an initialized state.
+ * </ul>
+ */
+// CodeNarc incorrectly thinks ".class" is unnecessary in @RunWith
+@SuppressWarnings('UnnecessaryDotClass')
+@RunWith(SpockRunner.class)
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy:58]: Why is it necessary? I would expect the `.class` to be implicit now that this is groovy.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+  /** Override to set config before the agent is installed */
+  protected void configurePreAgent() {}
+
+  def setup() {
+    configureLoggingLevels()
+
+    assert TEST_TRACER.activeSpan() == null: "Span is active before test has started: " + TEST_TRACER.activeSpan()
+
+    configurePreAgent()
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy:145]: Seems odd to call this again in `setup()` after the agent is initialized.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+    TEST_TRACER.flush()
+    TEST_WRITER.start()
+
+    new MockUtil().attachMock(STATS_D_CLIENT, this)
+  }
+
+  void cleanup() {
+    TEST_TRACER.flush()
+    new MockUtil().attachMock(STATS_D_CLIENT, this)
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy:157]: Why call this again here?

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+  }
+
+  /** Override to clean up things after the agent is removed */
+  protected void cleanupAfterAgent() {}
+
+  def cleanupSpec() {
+    TEST_TRACER?.close()
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy:164]: Should the GlobalTracer be reset with a no-op instance before/after this is closed?

File: dd-java-agent/testing/src/test/groovy/context/FieldBackedProviderTest.groovy (hunk 3 above)
Comment (randomanderson) [reply dd-java-agent/testing/src/test/groovy/context/FieldBackedProviderTest.groovy:43]: The previous code had a static instance named `TEST_LISTENER` that implemented `AgentBuilder.Listener` and called `onInstrumentationError` and `shouldTransformClass` on subclasses of `AgentTestRunner`.

With this PR, `AgentTestRunner` implements `AgentBuilder.Listener` directly:
shouldTransformClass() -> onDiscovery()
onInstrumentationError() -> onError()

`TEST_LISTENER` previously threw  `AbortTransfomationException` when `shouldTransformClass` returned false.

The test for `ContextTestInstrumentation.getName() + '$Incorrect'` was moved to `AgentTestRunner`.  The incorrect api usage instrumentation errors were being picked up by more that just this particular test because of the change in the ordering of the test lifecycle.  If that wasn't the case, `onInstrumentationError` would have just been changed to `onError`.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy (hunk 5 above)
Comment (randomanderson) [reply dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy:145]: Config is reset before each test.  If a property was set in `configurePreAgent()`, it would be gone by the second test.  The old code did this as well it was just less clear because the `@Before`s were scattered about the class.

I'll add a comment

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy (hunk 4 above)
Comment (randomanderson) [reply dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy:58]: I'm not sure but compilation fails without it.

```
/dd-java-agent/testing/build/tmp/compileGroovy/groovy-java-stubs/datadog/trace/agent/test/AgentTestRunner.java:12: error: cannot find symbol
@java.lang.SuppressWarnings(value="UnnecessaryDotClass") @org.junit.runner.RunWith(value=SpockRunner) public abstract class AgentTestRunner
                                                                                         ^
  symbol: variable SpockRunner
1 error
startup failed:
Compilation failed; see the compiler error output for details.

1 error


FAILURE: Build failed with an exception.
```

File: dd-trace-api/src/main/java/datadog/trace/api/GlobalTracer.java (hunk 2 above)
Comment (randomanderson) [reply dd-trace-api/src/main/java/datadog/trace/api/GlobalTracer.java:36]: No, and removing them would break tests.

The callbacks are added with the initialization of the logging framework specific `MDC` class (with `LogContextScopeListener.add()`).  This only happens once per module.  Everything worked fine previously because `CoreTracer` was static and only created once per module.

With the changes in this PR, `CoreTracer` is created once per test class but the `MDC` initialization still only happens once per module.  Those callbacks need to be called on every `CoreTracer` instance.

In the future, if we have classloader isolation between test classes in the same module, the `MDC` class would be initialized multiple times, but the static `installationCallbacks` field would also be initialized multiple times.

**tl;dr**: the lifecycle of installationCallbacks needs to match the lifecycle of the MDC subclass which is static for now.

File: dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy (hunk 6 above)
Comment (randomanderson) [reply dd-java-agent/testing/src/main/groovy/datadog/trace/agent/test/AgentTestRunner.groovy:157]: this is a bug, should be `detach`

**Review by tylerbenson — APPROVED** [review]

**Review by lpriima — APPROVED** [review]

File: dd-trace-api/src/main/java/datadog/trace/api/GlobalTracer.java [hunk 8]
```diff
@@ -31,50 +32,52 @@ public boolean addTraceInterceptor(TraceInterceptor traceInterceptor) {
...
+      if (provider == NO_OP) {
+        provider = p;
+        for (Callback callback : installationCallbacks) {
           callback.installed(p);
         }
       }
     }
   }
 
+  public static void forceRegister(Tracer tracer) {
```
Comment (lpriima) [inline dd-trace-api/src/main/java/datadog/trace/api/GlobalTracer.java:54]: This is not suppose public api, right ? Maybe comment
